packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol

```diff
@@ -1173,46 +997,61 @@ contract VotingV2 is Staker, OracleInterface, OracleAncillaryInterface, OracleGo
-        // within the price request struct to make the next entry into this method a no-op for this request.
-        _removeRequestFromPendingPriceRequests(priceRequest.pendingRequestIndex);
-
-        priceRequest.pendingRequestIndex = UINT64_MAX;
-        emit PriceResolved(
-            priceRequest.lastVotingRound,
-            priceRequest.priceRequestIndex,
-            priceRequest.identifier,
-            priceRequest.time,
-            priceRequest.ancillaryData,
-            resolvedPrice
-        );
-        return true;
+    function _resolveResolvablePriceRequests(uint256 maxTraversals) private {
+        uint32 currentRoundId = uint32(getCurrentRoundId());
+
+        uint256 index = rounds[currentRoundId].resolvableIndex;
+        uint256 requestsTraversed = 0;
+        while (index < pendingPriceRequestsIds.length && requestsTraversed <= maxTraversals) {
+            ++requestsTraversed;
+
+            PriceRequest storage request = priceRequests[pendingPriceRequestsIds[index]];
+            if (request.lastVotingRound >= currentRoundId) {
+                ++index;
+                continue;
+            }
+            VoteInstance storage voteInstance = request.voteInstances[request.lastVotingRound];
+            (bool isResolvable, int256 resolvedPrice) =
+                voteInstance.resultComputation.getResolvedPrice(_computeGat(request.lastVotingRound));
+            if (!isResolvable) {
+                if (request.lastVotingRound < currentRoundId) {
+                    request.rollCount += currentRoundId - request.lastVotingRound;
+                    request.lastVotingRound = currentRoundId;
+                    emit PriceRequestRolled(request.identifier, request.time, request.ancillaryData, request.rollCount);
+                    if (request.rollCount == deleteAfterRollCount && !request.isGovernance) {
+                        emit PriceRequestDeleted(request.identifier, request.time, request.ancillaryData);
+
+                        delete priceRequests[pendingPriceRequestsIds[index]];
+                        _removeRequestFromPendingPriceRequestsIds(request.pendingRequestIndex);
+                    }
+                } else ++index;
+                continue;
+            }
+
+            request.resolvedRequestIndex = SafeCast.toUint64(resolvedPriceRequestIds.length);
+            resolvedPriceRequestIds.push(pendingPriceRequestsIds[index]);
+
+            _removeRequestFromPendingPriceRequestsIds(request.pendingRequestIndex);
+            request.pendingRequestIndex = UINT64_MAX;
+
+            emit PriceResolved(
+                request.lastVotingRound,
+                request.resolvedRequestIndex,
+                request.identifier,
+                request.time,
+                request.ancillaryData,
+                resolvedPrice
+            );
+        }
+        rounds[currentRoundId].resolvableIndex = index;
```

Comment:
Why do we only set this when a request is resolved? Shouldn't we be setting it when the request is incremented? Or at the end after all the incrementing is done?

The way this is done here, I would think a huge number of spam rolls might cause this loop to never be able to mark its progress since there would be no resolved elements.

Any reason we couldn't set it outside the loop at whatever the ending value of index is?

Reply:
I realize now that this may have already been outside of the loop -- sorry the brackets are hard to track in the diff.

Reply:
yup! it's at the very end